File: 2020/javascript/day2/answer.js
const fs = require("fs")

const input = fs
    .readFileSync("input", { encoding: "utf8", flag: "r" })
    .split("\n")

function parse(line) {
    const regex = /(\d+)-(\d+) (\w): (\w+)/

    // return [min/pos, max/pos, char, password]
    return regex.exec(line).slice(1)
}

function validate_part1([min, max, char, password]) {
    const charCount = [...password].filter((letter) => letter == char).length

    return charCount >= min && charCount <= max
}

function validate_part2([pos1, pos2, char, password]) {
    return (password[pos1 - 1] == char) ^ (password[pos2 - 1] == char)
}

function validPasswordCount(input, validate) {
    return input
        .map(parse)
        .map(validate)
        .filter((x) => x).length
}

console.log(`Part 1: ${validPasswordCount(input, validate_part1)}`)
console.log(`Part 2: ${validPasswordCount(input, validate_part2)}`)

File: 2020/python/day2/answer.py
import re

def parse(line):
  parsed = re.search('(\d+)-(\d+) (\w): (\w+)', line)
  return (int(parsed.group(1)), int(parsed.group(2)), parsed.group(3), parsed.group(4))

def validate_part1(min, max, char, password):
  return min <= password.count(char) <= max

def validate_part2(pos1, pos2, char, password):
  return (password[pos1 - 1] == char) ^ (password[pos2 - 1] == char)

def valid_password_count(input, validate):
  return len([True for x in input if validate(*x) ])

with open('input') as f:
  input = [ parse(x) for x in f.read().split('\n')]
  
  print(f'Part 1: {valid_password_count(input, validate_part1)}')
  print(f'Part 2: {valid_password_count(input, validate_part2)}')
  
